scipy/linalg/src/_batched_linalg.h

```diff
@@ -0,0 +1,305 @@
+#pragma once
+
+#include "Python.h"
+#include <limits>
+#include <cassert>
+#include "numpy/arrayobject.h"
+#include "npy_cblas.h"
+#include "_lapack_trampolines.h"
+#include "_npymath.h"
+
+
+using namespace _numpymath;
+
+
+// parroted from sqrtm
+// XXX can probably be replaced by ?copy from BLAS
+template<typename T>
+inline void
+swap_cf(T* src, T* dst, const Py_ssize_t r, const Py_ssize_t c, const Py_ssize_t n)
+{
+    Py_ssize_t i, j, ith_row, r2, c2;
+    T *bb = dst;
+    T *aa = src;
+    if (r < 16) {
+        for (j = 0; j < c; j++)
+        {
+            ith_row = 0;
+            for (i = 0; i < r; i++) {
+                bb[ith_row] = aa[i];
+                ith_row += n;
+            }
+            aa += n;
+            bb += 1;
+        }
+    } else {
+        // If tall
+        if (r > c)
+        {
+            r2 = r/2;
+            swap_cf(src, dst, r2, c, n);
+            swap_cf(src + r2, dst+(r2)*n, r-r2, c, n);
+        } else {  // Nope
+            c2 = c/2;
+            swap_cf(src, dst, r, c2, n);
+            swap_cf(src+(c2)*n, dst+c2, r, c-c2, n);
+        }
+    }
+}
+
+
+/*
+ * Visit (m, n)-shaped slices of a ndim>=3 array in C order.
+ * Copy each slice into a buffer, make the buffer F-ordered
+ * for LAPACK.
+ */
+struct iter_data_t
```

Comment:
I don't get your comment about not using OOP. This is pretty much a class with a single method. Instead of loop optimizations now you are converting it into a Python generator and `yield`ing a slice here if I can read C++ correctly.